src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,438 @@
+
+    if (!joint)
+    {
+      std::stringstream message;
+      message << "Joint " << jointName << " (index: " << i << ")"
+        << " does not exist in metaSkeleton.";
+      throw std::invalid_argument{message.str()};
+    }
+    auto index = metaSkeleton->getIndexOf(joint);
+    assert(index != dart::dynamics::INVALID_INDEX);
+
+    rosJointToMetaSkeletonJoint.emplace(std::make_pair(i, index));
+
+  }
+
+  // Extract the first waypoint to infer the dimensionality of the trajectory.
+  Eigen::VectorXd currPosition, currVelocity, currAcceleration;
+  extractJointTrajectoryPoint(jointTrajectory, 0, numControlledJoints,
+    &currPosition, true, &currVelocity, false, &currAcceleration, false,
+    rosJointToMetaSkeletonJoint);
+
+  const auto& firstWaypoint = jointTrajectory.points.front();
+  auto currTimeFromStart = firstWaypoint.time_from_start.toSec();
+
+  const auto isPositionRequired = true;
+  const auto isVelocityRequired = (currVelocity.size() != 0);
+  const auto isAccelerationRequired = (currAcceleration.size() != 0);
+  if (isAccelerationRequired && !isVelocityRequired)
+  {
+    throw std::invalid_argument{
+      "Velocity is required since acceleration is specified."};
+  }
+
+  int numCoefficients;
+  if (isAccelerationRequired)
+    numCoefficients = 6; // quintic
+  else if (isVelocityRequired)
+    numCoefficients = 4; // cubic
+  else
+    numCoefficients = 2; // linear
+
+  // Convert the ROS trajectory message to an Aikido spline.
+  std::unique_ptr<SplineTrajectory> trajectory{new SplineTrajectory{space}};
+  auto currState = space->createState();
+
+  const auto& waypoints = jointTrajectory.points;
+  for (size_t iwaypoint = 1; iwaypoint < waypoints.size(); ++iwaypoint)
+  {
+    Eigen::VectorXd nextPosition, nextVelocity, nextAcceleration;
+    extractJointTrajectoryPoint(jointTrajectory, iwaypoint, numControlledJoints,
+      &nextPosition, isPositionRequired,
+      &nextVelocity, isVelocityRequired,
+      &nextAcceleration, isAccelerationRequired,
+      rosJointToMetaSkeletonJoint);
+
+    // Compute spline coefficients for this polynomial segment.
+    const auto nextTimeFromStart = waypoints[iwaypoint].time_from_start.toSec();
+    const auto segmentDuration = nextTimeFromStart - currTimeFromStart;
+    const auto segmentCoefficients = fitPolynomial(
+      0., Eigen::VectorXd::Zero(numControlledJoints), currVelocity, currAcceleration,
+      segmentDuration, nextPosition - currPosition, nextVelocity, nextAcceleration,
+      numCoefficients);
+
+    // Add a segment to the trajectory.
+    space->convertPositionsToState(currPosition, currState);
+    trajectory->addSegment(segmentCoefficients, segmentDuration, currState);
+
+    // Advance to the next segment.
+    currPosition = nextPosition;
+    currVelocity = nextVelocity;
+    currAcceleration = nextAcceleration;
+    currTimeFromStart = nextTimeFromStart;
+  }
+
+  return std::move(trajectory);
```

Comment:
`std::move( )` is unnecessary. Related [SO answer](http://stackoverflow.com/a/11817902).